File: src/17장/17-4.py
def upper_bound(s, e, d,L):
    while(e - s > 0):
        m = (s+e)//2
        if(L[m] <= d):
            s = m+1
        else:
            e = m
    return e
    
def lower_bound(s, e, d,L):
    while(e - s > 0):
        m = (s+e)//2
        if(L[m] < d):
            s = m+1
        else:
            e = m
    return e

up=[]
down=[]
result=[0]*500001
n,h=map(int,input().split())
for i in range(n):
  obstacle=int(input())
  if i%2==1: up.append(obstacle)
  else :down.append(obstacle)

up.sort()
down.sort()

answer=0
mx=2147483647

for i in range(1,h+1):
  idxd=lower_bound(0,len(down),i,down)
  idxu=lower_bound(0,len(up),h-i+1,up)
  result[i]=n//2-idxd + n//2-idxu
  mx=min(mx,result[i])

for i in range(1,h+1):
  if result[i]==mx:answer+=1

print(mx,answer)

File: src/17장/17-4.java
import java.util.*;
import java.io.*;

public class Main {
	static int N,H,up[], down[];
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		N=Integer.parseInt(st.nextToken());
		H=Integer.parseInt(st.nextToken());
		up=new int[N/2];
		down=new int[N/2];
		int upIdx=0;
		int downIdx=0;
		int mx=Integer.MAX_VALUE;
		int ans=1;
		for(int i=0; i<N; i++) {
			int num=Integer.parseInt(br.readLine());
			if(i%2==0) down[downIdx++]=num;
			else up[upIdx++]=num;
		}
		Arrays.sort(up);
		Arrays.sort(down);
		for(int i=1; i<=H; i++) { 
			int idxd=binarySearch(0,N/2-1 , i, down); 
			int idxu=binarySearch(0,N/2-1 , H-i+1, up); 
			if(mx>idxd+idxu) {
				mx=idxd+idxu;
				ans=1;
			}
			else if(mx==idxd+idxu) {
				ans++;
			}
		}
		System.out.print(mx+" "+ans);
	}
	private static int binarySearch(int left, int right, int height, int[] arr) {
		int min=Integer.MAX_VALUE;
		while(left<=right) {
			int mid = (left+right)/2;
			if(arr[mid]>=height) { 
				min=Math.min(min, mid);
				right=mid-1; 
			}
			else left=mid+1;
		}
		return min==Integer.MAX_VALUE? 0 : (N/2)-min;
	}
}